onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.h

```diff
@@ -69,6 +75,32 @@ QLinearOpType GetQLinearOpType(const onnxruntime::Node& node);
 // Such as QLinearConv, QLinearMatMul, QLinearAdd, ...
 bool IsQLinearBinaryOp(QLinearOpType qlinear_op_type);
 
+// Check if a qlinear binary op has valid inputs
+bool HasValidBinaryOpQuantizedInputs(const Node& node);
+// Check if a qlinear op has valid scales for given indices
+bool HasValidQuantizationScales(const InitializerMap& initializers, const Node& node,
+                                const std::vector<size_t>& indices);
+// Check if a qlinear op has valid zero points for given indices
+bool HasValidQuantizationZeroPoints(const InitializerMap& initializers, const Node& node,
+                                    const std::vector<size_t>& indices);
+
+// Get initialize tensort float/int32/int64 data without unpacking
+// TODO, move to ort framework
+const float* GetTensorFloatData(const ONNX_NAMESPACE::TensorProto& tensor);
+const int32_t* GetTensorInt32Data(const ONNX_NAMESPACE::TensorProto& tensor);
+const int64_t* GetTensorInt64Data(const ONNX_NAMESPACE::TensorProto& tensor);
+
```

Comment:
Do we need this or can UnpackTensor in tensorprotoutils be used instead to save some binary size?

Reply:
Using GetTensorTYPEData will return the initiailizer data directly instead of copy them to a new piece of memory
Not sure with extra code (to allocate an array and a size_t) then unpack (and check result) will actually reduce binary size

---
In reply to: [520403511](https://github.com/microsoft/onnxruntime/pull/5746#discussion_r520403511) [](ancestors = 520403511)